ksql-engine/src/test/java/io/confluent/ksql/EndToEndEngineTest.java

```diff
@@ -0,0 +1,531 @@
+import org.apache.kafka.streams.test.ConsumerRecordFactory;
+import org.apache.kafka.streams.test.OutputVerifier;
+import org.apache.kafka.test.TestUtils;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Ignore;
+import org.junit.Test;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+import java.util.stream.Collectors;
+
+import io.confluent.kafka.schemaregistry.client.MockSchemaRegistryClient;
+import io.confluent.ksql.metastore.MetaStore;
+import io.confluent.ksql.metastore.MetaStoreImpl;
+import io.confluent.ksql.util.FakeKafkaTopicClient;
+import io.confluent.ksql.util.KsqlConfig;
+import io.confluent.ksql.util.QueryMetadata;
+
+class EndToEndEngineTest {
+  protected interface SerdeSupplier<T> {
+    Serializer<T> getSerializer(SchemaRegistryClient schemaRegistryClient);
+    Deserializer<T> getDeserializer(SchemaRegistryClient schemaRegistryClient);
+  }
+
+  static class StringSerdeSupplier implements SerdeSupplier<String> {
+    @Override
+    public Serializer<String> getSerializer(final SchemaRegistryClient schemaRegistryClient) {
+      return Serdes.String().serializer();
+    }
+
+    @Override
+    public Deserializer<String> getDeserializer(final SchemaRegistryClient schemaRegistryClient) {
+      return Serdes.String().deserializer();
+    }
+  }
+
+  static class AvroSerdeSupplier implements SerdeSupplier {
+    @Override
+    public Serializer getSerializer(final SchemaRegistryClient schemaRegistryClient) {
+      return new KafkaAvroSerializer(schemaRegistryClient);
+    }
+
+    @Override
+    public Deserializer getDeserializer(final SchemaRegistryClient schemaRegistryClient) {
+      return new KafkaAvroDeserializer(schemaRegistryClient);
+    }
+  }
+
+  static class ValueSpecAvroDeserializer implements Deserializer<Object> {
+    private final SchemaRegistryClient schemaRegistryClient;
+    private final KafkaAvroDeserializer avroDeserializer;
+
+    public ValueSpecAvroDeserializer(final SchemaRegistryClient schemaRegistryClient) {
+      this.schemaRegistryClient = schemaRegistryClient;
+      this.avroDeserializer = new KafkaAvroDeserializer(schemaRegistryClient);
+    }
+
+    @Override
+    public void close() {
+    }
+
+    @Override
+    public void configure(final Map<String, ?> properties, final boolean b) {
+    }
+
+    @Override
+    public Object deserialize(final String topicName, final byte[] data) {
+      Object avroObject = avroDeserializer.deserialize(topicName, data);
+      String schemaString;
+      try {
+        schemaString = schemaRegistryClient.getLatestSchemaMetadata(topicName + "-value").getSchema();
+      } catch (Exception e) {
+        throw new RuntimeException(e);
+      }
+      return avroToValueSpec(
```

Comment:
nit: you could move this up into the `try{..}catch` and then `schemaString` can be initialized where it is defined

Reply:
I'd prefer to minimize the code under the catch-all to avoid rewrapping exceptions as much as possible.